src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -895,21 +956,51 @@ private static unsafe void Write(TextWriter output, char* input, int inputLength
             return encoding;
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private int DoesAsciiNeedEncoding(byte value)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private unsafe void InitializeAsciiCache()
         {
-            Debug.Assert(value <= 0x7F);
+            Debug.Assert(!_isAsciiCacheInitialized);
+            Debug.Assert(_asciiNeedsEscaping.Length == 0x80);
+
+            _isAsciiCacheInitialized = true;
+            bool[] asciiNeedsEscaping = _asciiNeedsEscaping;
+
+#if NETCOREAPP
+            if (Ssse3.IsSupported)
+            {
+                sbyte* tmp = stackalloc sbyte[Vector128<sbyte>.Count];
+                Sse2.Store(tmp, Vector128<sbyte>.Zero);
+
+                for (int i = 0; i < asciiNeedsEscaping.Length; i++)
+                {
+                    bool willEncode = WillEncode(i);
+                    asciiNeedsEscaping[i] = willEncode;
 
-            int needsEscaping = _asciiNeedsEscaping[value];
+                    if (willEncode)
+                    {
+                        int highNibble = i >> 4;
+                        int lowNibble = i & 0xF;
 
-            Debug.Assert(needsEscaping == 0 || needsEscaping == 1 || needsEscaping == -1);
+                        tmp[lowNibble] |= (sbyte)(1 << highNibble);
+                    }
+                }
 
-            if (needsEscaping == 0)
+                _bitMaskLookupAsciiNeedsEscaping = Sse2.LoadVector128(tmp);
+                return;
+            }
+#endif
+            for (int i = 0; i < asciiNeedsEscaping.Length; i++)
             {
-                needsEscaping = WillEncode(value) ? 1 : -1;
-                _asciiNeedsEscaping[value] = needsEscaping;
+                asciiNeedsEscaping[i] = WillEncode(i);
             }
-            return needsEscaping;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool DoesAsciiNeedEncoding(uint value)
+        {
+            Debug.Assert(value <= 0x7F);
+
+            return _asciiNeedsEscaping[value];
```

Comment:
Not really related to this PR, but just for fun, if you wanted to eke out even higher performance:

```cs
public unsafe class SomeEncoder
{
    private AsciiNeedsEscapingData _doesAsciiNeedEscaping;

    private bool DoesAsciiNeedEscaping(uint value)
    {
        return _doesAsciiNeedEscaping.Data[value];
    }

    private struct AsciiNeedsEscapingData
    {
        internal fixed bool Data[128];
    }
}
```

This will suppress the indexer bounds checks while accessing the fixed-size buffer.

Reply:
Thanks -- this is really cool 👍 

I'll incorporate this, as it's quite some reduction in generated code.